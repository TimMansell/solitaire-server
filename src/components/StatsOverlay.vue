<template>
  <GameOverlay alt center-content show-logo data-test="stats-overlay">
    <template #title>
      Stats
    </template>
    <template #msg>
      <p>Showing stats for completed games:</p>

      <Table
        :headings="['Played', 'Won', 'Lost', 'Gave Up']"
        :items="stats"
        :placeholder-rows="2"
      />
    </template>
    <template #buttons>
      <Button @click="closeStats" data-test="close-stats-btn">
        Close
      </Button>
    </template>
  </GameOverlay>
</template>

<script>
import numeral from 'numeral';
import { mapActions, mapGetters } from 'vuex';
import GameOverlay from '@/components/GameOverlay.vue';
import Button from '@/components/Button.vue';
import Table from '@/components/Table.vue';

export const calcPercent = (value) => numeral(value).format('0.00%');

export const calcNumber = (value) => numeral(value).format('0,0');

export const calcStats = ({ completed, won, lost }) => {
  const abandoned = completed - won - lost;

  const completedCount = calcNumber(completed);
  const wonCount = calcNumber(won);
  const lostCount = calcNumber(lost);
  const abandonedCount = calcNumber(abandoned);

  const wonPercent = calcPercent(won / completed);
  const lostPercent = calcPercent(lost / completed);
  const abandonedPercent = calcPercent(abandoned / completed);

  const stats = [
    [completedCount, wonCount, lostCount, abandonedCount],
    ['', wonPercent, lostPercent, abandonedPercent],
  ];

  if (completed >= 0) {
    return stats;
  }

  return [];
};

export default {
  name: 'GameLost',
  components: {
    GameOverlay,
    Button,
    Table,
  },
  computed: {
    ...mapGetters(['fullStats', 'isLoading']),
    stats() {
      const { fullStats } = this;
      const stats = calcStats(fullStats);

      return stats;
    },
  },
  methods: {
    ...mapActions(['toggleStats', 'clearFullStats']),
    closeStats() {
      this.toggleStats();
    },
  },
  destroyed() {
    this.clearFullStats();
  },
};
</script>
