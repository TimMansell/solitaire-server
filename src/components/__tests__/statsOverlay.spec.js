import { shallowMount } from '@vue/test-utils';
import StatsOverlay, {
  calcPercent,
  calcStats,
} from '@/components/StatsOverlay.vue';

const mocks = {
  $store: { dispatch: jest.fn() },
};

const fullStats = {
  completed: 9,
  won: 2,
  lost: 4,
};

const computed = {
  fullStats: () => ({
    ...fullStats,
  }),
};

describe('StatsOverlay.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(StatsOverlay, {
      computed,
      mocks,
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('should calculate correct % values', () => {
    const result = calcPercent(1 / 2);

    expect(result).toBe('50.00%');
  });

  it('should calculate correct stats', () => {
    const result = calcStats(fullStats);

    expect(result).toStrictEqual([
      ['9', '2', '4', '3'],
      ['', '22.22%', '44.44%', '33.33%'],
    ]);
  });
});
