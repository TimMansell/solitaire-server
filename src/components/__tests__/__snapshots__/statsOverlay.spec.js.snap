// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsOverlay.vue matches snapshot 1`] = `
<gameoverlay-stub centercontent="true" showlogo="true" visible="true" data-test="stats-overlay">
  Your Stats
  <table-stub headings="Played,Won,Lost,Gave Up" items="9,2,4,3,,22.22%,44.44%,33.33%" placeholderrows="2"></table-stub>
  <h1>Global Stats</h1>
  <table-stub headings="Played,Won,Lost,Gave Up" items="9,2,4,3,,22.22%,44.44%,33.33%" placeholderrows="2"></table-stub>
  <button-stub type="default" size="md" data-test="close-stats-btn">
    Close
  </button-stub>
</gameoverlay-stub>
`;
